File: Algorithms/1696_Jump_Game_VI/Java/JumpGameVISolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int maxResult(int[] nums, int k) {
        if (nums.length == 1) return nums[0];
        int[] dp = new int[nums.length];
        Deque<Integer> queue = new ArrayDeque<>();
        dp[0] = nums[0];
        queue.add(0);
        for (int i = 1; i < nums.length; i++) {
            dp[i] = nums[i] + dp[queue.peekFirst()];
            while (!queue.isEmpty() && dp[i] >= dp[queue.getLast()]) queue.pollLast();
            while (!queue.isEmpty() && i - queue.getFirst() >= k) queue.pollFirst();
            queue.add(i);
        }
        return dp[dp.length - 1];
    }
}

File: Algorithms/1696_Jump_Game_VI/Python/Jump_Game_VI_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def maxResult(self, nums, k):
        length = len(nums)
        if length == 1: return nums[0]

        dp = [0 for _ in range(length)]
        queue = [0]
        dp[0] = nums[0]

        for i in range(1, length):
            dp[i] = nums[i] + dp[queue[0]]

            while queue and dp[i] >= queue[-1]:
                queue.pop()

            while queue and i - len(queue) >= k:
                queue.pop(0)

            queue.append(i)

        return dp[-1]






